services/aster-nix/src/fs/exfat/inode.rs

```diff
@@ -0,0 +1,1646 @@
+            let file_size = inner.size;
+            let start = file_size.min(offset);
+            let end = file_size.min(offset + buf.len());
+            (start, end - start)
+        };
+        inner
+            .page_cache
+            .pages()
+            .read_bytes(read_off, &mut buf[..read_len])?;
+
+        inner.upgrade().atime = DosTimestamp::now()?;
+        Ok(read_len)
+    }
+
+    // The offset and the length of buffer must be multiples of the block size.
+    fn read_direct_at(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
+        let inner = self.inner.upread();
+        if inner.inode_type.is_directory() {
+            return_errno!(Errno::EISDIR)
+        }
+        if !is_block_aligned(offset) || !is_block_aligned(buf.len()) {
+            return_errno_with_message!(Errno::EINVAL, "not block-aligned");
+        }
+
+        let sector_size = inner.fs().sector_size();
+
+        let (read_off, read_len) = {
+            let file_size = inner.size;
+            let start = file_size.min(offset).align_down(sector_size);
+            let end = file_size.min(offset + buf.len()).align_down(sector_size);
+            (start, end - start)
+        };
+
+        inner
+            .page_cache
+            .pages()
+            .decommit(read_off..read_off + read_len)?;
+
+        let mut buf_offset = 0;
+        let frame = VmAllocOptions::new(1).uninit(true).alloc_single().unwrap();
+
+        for bid in BlockId::from_offset(read_off)..BlockId::from_offset(read_off + read_len) {
+            //TODO: performance needs to be improved.
+            let frame = VmAllocOptions::new(1).uninit(true).alloc_single().unwrap();
+            inner.fs().block_device().read_block_sync(bid, &frame)?;
+
+            frame.read_bytes(0, &mut buf[buf_offset..buf_offset + BLOCK_SIZE])?;
+            buf_offset += BLOCK_SIZE;
+        }
+
+        inner.upgrade().atime = DosTimestamp::now()?;
+        Ok(read_len)
+    }
+
+    fn write_at(&self, offset: usize, buf: &[u8]) -> Result<usize> {
+        //We need to obtain the fs lock to resize the file.
+        {
+            let mut inner = self.inner.write();
+            if inner.inode_type.is_directory() {
+                return_errno!(Errno::EISDIR)
+            }
+
+            let file_size = inner.size;
+            let file_allocated_size = inner.size_allocated;
+            let new_size = offset + buf.len();
+            let fs = inner.fs();
+            let fs_guard = fs.lock();
+            if new_size > file_size {
+                if new_size > file_allocated_size {
+                    inner.resize(new_size, &fs_guard)?;
+                }
+                inner.page_cache.pages().resize(new_size)?;
+                inner.size = new_size;
+            }
+        }
+
+        //locks released here, so that file write can be parallized.
+
+        let inner = self.inner.upread();
+        inner.page_cache.pages().write_bytes(offset, buf)?;
```

Comment:
The `write_bytes` of PageCache method will load the page first if the `PageCacheBackend` have enough pages.

So If we update the size of the inode first, and call the `write_bytes` method, it must do an unnecessary `read_page`,  and the data on the page may be ugly(which means the data on that page can not ensure zero).

In the implementation of Ramfs and Ext2, we always write data into the pagecache first and then update the size of the inode.
```rust
        let self_inode = self.0.upread();
        let Some(page_cache) = self_inode.inner.as_file() else {
            return_errno_with_message!(Errno::EISDIR, "write is not supported");
        };
        let file_size = self_inode.metadata.size;
        let new_size = offset + buf.len();
        let should_expand_size = new_size > file_size;
        if should_expand_size {
            page_cache.pages().resize(new_size)?;
        }
        page_cache.pages().write_bytes(offset, buf)?;
        if should_expand_size {
            // Turn the read guard into a write guard without releasing the lock.
            let mut self_inode = self_inode.upgrade();
            self_inode.resize(new_size);
        }
```